frontend/javascripts/oxalis/model/volumetracing/volumelayer.js

```diff
@@ -277,79 +208,77 @@ class VolumeLayer {
     // area if you consider narrow shapes.
     // Also, it will be very clear where to start the filling
     // algorithm.
-    this.drawOutlineVoxels((x, y) => setMap(x, y, false), mode);
+    this.drawOutlineVoxels((x, y) => setMap(x, y, 0), mode);
     this.fillOutsideArea(map, width, height);
     this.drawOutlineVoxels(setMap, mode);
 
-    const numberOfSlices = getNumberOfSlicesForResolution(activeResolution, this.plane);
-    const thirdDimensionIndex = Dimensions.thirdDimensionForPlane(this.plane);
-
-    const iterator = new VoxelIterator(
+    const buffer2D = new VoxelBuffer2D(
       map,
       width,
       height,
       minCoord2d,
       this.get3DCoordinate.bind(this),
-      thirdDimensionIndex,
-      numberOfSlices,
+      this.getFast3DCoordinateFunction(),
     );
-    return iterator;
+    return buffer2D;
   }
 
-  getCircleVoxelIterator(
-    position: Vector3,
-    activeResolution: Vector3,
-    boundings?: ?BoundingBoxType,
-  ): VoxelIterator {
+  getCircleVoxelBuffer2D(position: Vector3): VoxelBuffer2D {
     const state = Store.getState();
     const { brushSize } = state.userConfiguration;
+    const dimIndices = Dimensions.getIndices(this.plane);
 
-    const radius = Math.round(brushSize / 2);
-    const width = 2 * radius;
-    const height = 2 * radius;
+    const unzoomedRadius = Math.round(brushSize / 2);
+    const width = Math.floor((2 * unzoomedRadius) / this.activeResolution[dimIndices[0]]);
+    const height = Math.floor((2 * unzoomedRadius) / this.activeResolution[dimIndices[1]]);
 
-    const map = new Array(width);
-    for (let x = 0; x < width; x++) {
-      map[x] = new Array(height).fill(false);
-    }
-    const floatingCoord2d = this.get2DCoordinate(position);
-    const coord2d = [Math.floor(floatingCoord2d[0]), Math.floor(floatingCoord2d[1])];
-    const minCoord2d = [coord2d[0] - radius, coord2d[1] - radius];
+    const map = new Uint8Array(width * height).fill(0);
+
+    const floatingCoord2d = this.get2DCoordinate(
+      scaleGlobalPositionWithResolutionFloat(position, this.activeResolution),
+    );
+    const minCoord2d = [
+      Math.floor(floatingCoord2d[0] - width / 2),
+      Math.floor(floatingCoord2d[1] - height / 2),
+    ];
 
     // Use the baseVoxelFactors to scale the circle, otherwise it'll become an ellipse
     const [scaleX, scaleY] = this.get2DCoordinate(
       getBaseVoxelFactors(state.dataset.dataSource.scale),
     );
 
     const setMap = (x, y) => {
-      map[x][y] = true;
+      map[x * height + y] = 1;
     };
-    Drawing.fillCircle(radius, radius, radius, scaleX, scaleY, setMap);
-
-    const numberOfSlices = getNumberOfSlicesForResolution(activeResolution, this.plane);
-    const thirdDimensionIndex = Dimensions.thirdDimensionForPlane(this.plane);
+    Drawing.fillCircle(
+      Math.floor(unzoomedRadius / this.activeResolution[dimIndices[0]]),
+      Math.floor(unzoomedRadius / this.activeResolution[dimIndices[1]]),
+      unzoomedRadius,
```

Comment:
Yeah, it's quite a bit of a brain twister :see_no_evil: But the current code should be correct. The crucial part is that the scaleX and scaleY parameters for the `fillCircle` call are already adapted to take the resolution into account. The radius itself cannot be adapted here, since its a unary value (and the dimensions can be anisotropic).
GitHub does not show the critical context, which is why I'm pasting it here again:

```
    Drawing.fillCircle(
      Math.floor(unzoomedRadius / this.activeResolution[dimIndices[0]]),
      Math.floor(unzoomedRadius / this.activeResolution[dimIndices[1]]),
      unzoomedRadius,
      scaleX / this.activeResolution[dimIndices[0]], // <-- this effectively scales the unzoomedRadius to a zoomedRadius
      scaleY / this.activeResolution[dimIndices[1]],
      setMap,
    );
```

I'll add an additional comment to the code to make it clearer!

Reply:
That's what I suspected, thanks a lot for clarifying :)